### PR TITLE
Move GC blobs to root and re-generate GC / summary on GC state switch

### LINKED ISSUE #8515: Enable writing the GC data in summary at root
Follow up to https://github.com/microsoft/FluidFramework/pull/8389.
The new GC data format in summary is shipped dark and is behind a feature flag. Once this code saturates, we should enable writing the new format.

### LINKED ISSUE #8672: Stress: `fluid:telemetry:ContainerRuntime:GarbageCollector:gcUnknownOutboundRoute`
Looks of `fluid:telemetry:ContainerRuntime:GarbageCollector:gcUnknownOutboundRoute` errors in stress runs:
let buildid = "47789";
office_fluid_ffautomation_error  | where Data_buildId  == buildid
| project-reorder Event_Time, Data_eventName, Data_error, Data_message, Data_reason, Data_loaderVersion, Data_runtimeVersion, Data_stack, Data_fluidErrorCode

### PR DESCRIPTION
**Pull request by agarwal-navin**
Fixes #8672 and #8515.

As mentioned in the root-cause for #8672 [here](https://github.com/microsoft/FluidFramework/issues/8672#issuecomment-1009542430), when GC is disabled after it was previously enabled, the summary has to re-regenerate the data for all nodes so that their unreferenced state is cleared.
This PR adds the following changes:
- Moved writing the GC data from data stores to the root. This was a pending work item it makes finding out if a summary has GC state simpler and helps having all logic in one place (garbage collector).
- When the GarbageCollector loads, it checks if the summary has GC tree or not. If it does and GC is disabled, it sets a flag indicating that the GC state needs to be reset in the next summary.
- ContainerRuntime when generating summary, consults the above flag and if true, runs summary with `fullTree` as true.
- The flag is reset when the first summary is ack'd since we only want to regenerate the summary the very first time this happens.

Note that moving the GC blobs to root will have the following consequence:
- A forward compat change was made so that older versions can read the GC data at root and write at root here - https://github.com/microsoft/FluidFramework/pull/8389.
- But if a summarizer with this change generates a summary which is read by another summarizer running version < 0.53, it will have to regenerate the summary and GC data. 

PR that updates the reference snapshots used in snapshot tests - https://github.com/microsoft/FluidFrameworkTestData/pull/35

### CODE REVIEW

File: packages/runtime/container-runtime/src/garbageCollection.ts [hunk 1]
```diff
@@ -342,9 +344,10 @@ export class GarbageCollector implements IGarbageCollector {
         // Whether we are running in test mode. In this mode, unreferenced nodes are immediately deleted.
         this.testMode = this.mc.config.getBoolean(gcTestModeKey) ?? gcOptions.runGCInTestMode === true;
 
-        // If `writeDataAtRoot` GC option is true, we should write the GC data into the root of the summary tree. This
-        // GC option is used for testing only. It will be removed once we start writing GC data into root by default.
-        this._writeDataAtRoot = this.gcOptions.writeDataAtRoot === true;
+        // The GC state needs to be reset if the base snapshot contains GC tree and GC is disabled or it doesn't contain
+        // GC tree and GC is enabled.
+        const gcTreePresent = baseSnapshot?.trees[gcTreeKey] !== undefined;
+        this.initialStateNeedsReset = gcTreePresent ? !this.shouldRunGC : this.shouldRunGC;
```
Comment (markfields) [inline packages/runtime/container-runtime/src/garbageCollection.ts:350]: nit: use && or || operators not ternary operator?

File: packages/runtime/container-runtime/src/garbageCollection.ts [hunk 2]
```diff
@@ -221,13 +219,16 @@ export class GarbageCollector implements IGarbageCollector {
...
+     * 3. The GC version in the latest summary is different from the current GC version. This can happen if:
+     *    3.1. The summary this client loaded with has data from a different GC version.
+     *    3.2. This client's latest summary was updated from a snapshot that has a different GC version.
      */
-    public get hasGCVersionChanged(): boolean {
-        // The current version can differ from the latest summary version in two cases:
-        // 1. The summary this client loaded with has data from a different GC version.
-        // 2. This client's latest summary was updated from a snapshot that has a different GC version.
-        return this.shouldRunGC && this.latestSummaryGCVersion !== this.currentGCVersion;
+    public get summaryStateNeedsReset(): boolean {
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/garbageCollection.ts:229]: How does this get used in the 1st scenario?
>GC was enabled and is now disabled. The GC state needs to be removed and everything becomes referenced.

File: packages/test/test-end-to-end-tests/src/test/gc/gcContainerRuntimeCompat.spec.ts [hunk 3]
```diff
@@ -44,7 +41,7 @@ describeFullCompat("GC summary compatibility tests", (getTestObjectProvider) =>
         []);
     const runtimeOptions: IContainerRuntimeOptions = {
         summaryOptions: { disableSummaries: true },
-        gcOptions: { gcAllowed: true },
+        gcOptions: { gcAllowed: true, writeDataAtRoot: true },
```
Comment (tyler-cai-microsoft) [inline packages/test/test-end-to-end-tests/src/test/gc/gcContainerRuntimeCompat.spec.ts:44]: Aren't you removing `writeDataAtRoot` in the files above?

Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcContainerRuntimeCompat.spec.ts:44]: Yes, this is for compatibility. Older versions of runtime still expect this option to write the data at root.

File: packages/runtime/container-runtime/src/garbageCollection.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/garbageCollection.ts:229]: In the first scenario, `initialStateNeedsReset` will be true.

File: packages/test/test-end-to-end-tests/src/test/gc/gcContainerRuntimeCompat.spec.ts (hunk 3 above)
Comment (pleath) [reply packages/test/test-end-to-end-tests/src/test/gc/gcContainerRuntimeCompat.spec.ts:44]: ...or is this simulating an older version that would have had writeDataAtRoot defined?

File: packages/test/test-end-to-end-tests/src/test/gc/gcStateResetInSummaries.spec.ts [hunk 4]
```diff
@@ -228,23 +213,52 @@ describeFullCompat("GC state reset in summaries", (getTestObjectProvider) => {
...
         );
-        await validateGCDidNotRun(summarizerClient2);
+        // Validate that GC does not run and the summary is regenerated because GC was disabled.
+        await validateGCState(
+            summarizerClient2,
+            false /* shouldGCRun */,
+            true /* shouldRegenerateSummary */,
+        );
+
+        // Validate that GC does not run and the summary should be regenerated again. The summary is regenerated only
```
Comment (pleath) [inline packages/test/test-end-to-end-tests/src/test/gc/gcStateResetInSummaries.spec.ts:241]: Should this read, "summary should not be regenerated"?

File: packages/test/test-end-to-end-tests/src/test/gc/gcStateResetInSummaries.spec.ts [hunk 5]
```diff
@@ -228,23 +213,52 @@ describeFullCompat("GC state reset in summaries", (getTestObjectProvider) => {
...
+        );
+
+        // Load a new summarizer from the last summary with GC still disabled. Validate that summary is NOT regenerated
+        // by this client since the summary doesn't have any GC state.
+        const summarizerClient3 = await getNewSummarizer(
+            true /* disableGC */,
+            undefined /* gcAllowed */,
+            latestAckedSummary.summaryAck.contents.handle,
+        );
+        // Validate that GC does not run and the summary is regenerated because GC was disabled.
```
Comment (pleath) [inline packages/test/test-end-to-end-tests/src/test/gc/gcStateResetInSummaries.spec.ts:256]: Should this read "summary is not regenerated" also?

File: packages/test/test-end-to-end-tests/src/test/gc/gcStateResetInSummaries.spec.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcStateResetInSummaries.spec.ts:241]: yes, you are right. copy paste error :)

**Review by pleath — APPROVED** [review]

**Comment by agarwal-navin** [comment]
@pleath Could you please also take a look at the corresponding PR in test repo - https://github.com/microsoft/FluidFrameworkTestData/pull/35. That has to go in before I can merge this one.

File: packages/runtime/container-runtime/src/garbageCollection.ts [hunk 6]
```diff
@@ -221,13 +219,16 @@ export class GarbageCollector implements IGarbageCollector {
...
+        return this.initialStateNeedsReset ||
+            (this.shouldRunGC && this.latestSummaryGCVersion !== this.currentGCVersion);
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/garbageCollection.ts:231]: Wouldn't this allow an older client to overwrite a newer GC state?

**Pull request merged**